### DISCUSSION TITLE
modbusLAN working 100% after changing scan_interval

### DISCUSSION THREAD

**Discussion by TonyM1958** [StealthChesnut/HA-FoxESS-Modbus#115] [General]
I've been having problems with the modbusLAN ethernet port crashing. By a process of trial and error, I now seem to have this working reliably with Manager firmware 1.57.

There was a rumour that scanning the modbus server too often caused problems, so the change I made was to replace all instances of scan_interval in modbusLAN.yaml with !secret scan_interval with an initial setting of 30. I've now reduced this to 5 and it's not crashed on me for more than a week (it crashed randomly within 24 hours using the github file settings).

My (unproven) theory is that all the different scan_intervals might result in multiple timers running in HA to trigger scans and that the different intervals result in some sort of race condition where 2 sensor scans run at the same time, or overlap in some way that crashes the modbus server in the inverter...

Looking through the HA modbus documents, the default scan interval is 15 seconds. I have now removed all of the scan_interval settings in modbusLAN.yaml to see if this has any effect (good or bad!).

So, the question is: why are there so many different scan_interval settings in modbusLAN.yaml? For example, PV1_Voltage and PV1_Current are 30 seconds, but PV1_Power is 5 seconds. Does anyone know how the current values were decided and what the potential impact is of changing them?

**Comment by Daibutt** [comment]
Not sure I know the answer to your question, but sensors like power, voltage, current will vary over short intervals, so a low scan rate would potentially make sense for them.
Personally I'm not massively interested in voltage & current at 5/10 sec intervals so have changed these to 30/60/100 secs. 
If I have an issue where I'd want to look at these more closely I'll just reduce the the scan interval in my modbus yaml.

Temperatures, SoC, battery cycles, remaining kWh, battery voltage etc. won't be varying that quickly so these could be at 30sec+.
E.g. currently my SoC is set at 121 secs, for me more than adequate as SoC is not changing every 5 or 10 seconds.

**Reply by TonyM1958** [reply]
Hi @Daibutt, thanks for replying - its useful to talk some of these things through.

I understand the arguments for increasing or decreasing scan interval for different sensors: intuitively, a slower scan interval should reduce system load and database space, whilst faster scanning should increase system load and database space.

Its just my empirical evidence does not seem to support this. I think I'm seeing a timing issue, where having a variety of scan intervals causes the inverter modbus LAN server to crash but it quite happily copes with all of the sensors being scanned at the same interval, even down to 5 seconds. I haven't (yet) managed to pin this down to specific sensors or interval settings as it takes a few days after each change to see if there is a crash (or not).

I am not sure about the argument that a faster scan rate increases database space either as, looking at the History, recorder only stores new values when they change. Obviously, scanning noisy sensors at a high rate will still increase the storage.

So many variables - may be you (and others) just won't see a problem if you are setting the scan intervals to longer than 30 seconds anyway.

**Comment by TonyM1958** [comment]
Thought I might be going mad as it seemed rock solid, so I reverted to the main copy of modbusLAN.yaml from github, and the inverter crashed after a couple of hours with the front panel also unresponsive this time. Full off and back on to get it back up and running.

Now running with github scan intervals but modbus delay set to 1 to see how long it stays up with those settings...

**Comment by canton7** [comment]
> My (unproven) theory is that all the different scan_intervals might result in multiple timers running in HA to trigger scans and that the different intervals result in some sort of race condition where 2 sensor scans run at the same time, or overlap in some way that crashes the modbus server in the inverter...

Requests are made under a lock, so there's no way for two requests to be made at the same time.
